File: backend/data/extract.js
const ExcelJS = require("exceljs");
const AdmZip = require("adm-zip");
const fs = require("fs").promises;
const path = require("path");

// Config
const EXCEL_PATH = path.join(__dirname, "OnePager New MRP.xlsx");
const OUTPUT_IMAGE_DIR = path.join(__dirname, "../img");
const OUTPUT_JSON_PATH = path.join(__dirname, "../output.json");
const IMAGE_COLUMN = "A";
const EXTERNAL_IMAGE_DIR = "images";

// Ensure directories exist
async function ensureDirectories() {
  await fs.mkdir(OUTPUT_IMAGE_DIR, { recursive: true });
  await fs.mkdir(EXTERNAL_IMAGE_DIR, { recursive: true });
  console.log(`✅ Output directory ${OUTPUT_IMAGE_DIR} is writable`);
}

// Extract images from Excel (xl/media/)
async function extractImagesFromXlsx(xlsxPath) {
  const imageMap = {};
  try {
    const zip = new AdmZip(xlsxPath);
    const entries = zip.getEntries();
    const imageFiles = entries.filter((entry) =>
      entry.entryName.startsWith("xl/media/")
    );
    console.log(`Found ${imageFiles.length} image files in Excel`);

    imageFiles.forEach((entry, idx) => {
      try {
        const imgData = entry.getData();
        if (imgData.length > 0) {
          imageMap[idx + 1] = imgData;
          console.log(
            `Extracted image ${idx + 1}: ${entry.entryName}, size=${
              imgData.length
            } bytes`
          );
        } else {
          console.log(`⚠️ Empty image data for ${entry.entryName}`);
        }
      } catch (e) {
        console.log(
          `⚠️ Failed to extract image ${entry.entryName}: ${e.message}`
        );
      }
    });

    if (Object.keys(imageMap).length > 0) {
      const testImgData = Object.values(imageMap)[0];
      const testPath = path.join(OUTPUT_IMAGE_DIR, "test_image.png");
      await fs.writeFile(testPath, testImgData);
      console.log(`✅ Saved test image to ${testPath}`);
    }
    console.log(`✅ Total images extracted: ${Object.keys(imageMap).length}`);
  } catch (e) {
    console.log(`⚠️ Failed to process Excel as ZIP: ${e.message}`);
  }
  return imageMap;
}

// Clean and format price
function cleanPrice(price) {
  if (typeof price === "string") {
    const cleaned = price.replace("₹", "").replace(/,/g, "").trim();
    const value = parseFloat(cleaned);
    return isNaN(value) ? price : value;
  }
  return price;
}

// Save image to disk
async function saveImage(imgBytes, code, outputDir) {
  try {
    if (!imgBytes || imgBytes.length < 100) {
      console.log(
        `⚠️ Invalid image data for code ${code}: ${imgBytes.length} bytes`
      );
      return null;
    }
    const ext = "png"; // Adjust based on actual image type if needed
    const filename = `${code}.${ext}`;
    const filepath = path.join(outputDir, filename);
    await fs.writeFile(filepath, imgBytes);
    if (
      await fs
        .access(filepath)
        .then(() => true)
        .catch(() => false)
    ) {
      console.log(`✅ Saved image for code ${code}: ${filepath}`);
      return `./images/${filename}`;
    } else {
      console.log(`⚠️ Image file not found after saving: ${filepath}`);
      return null;
    }
  } catch (e) {
    console.log(`⚠️ Failed to save image for code ${code}: ${e.message}`);
    return null;
  }
}

// Main processing
async function main() {
  try {
    await ensureDirectories();

    // Load workbook
    const workbook = new ExcelJS.Workbook();
    await workbook.xlsx.readFile(EXCEL_PATH);

    // Check protection (limited support in exceljs)
    console.log(
      "⚠️ exceljs has limited protection detection. Check Excel manually for protection."
    );

    const imageMap = await extractImagesFromXlsx(EXCEL_PATH);
    const allData = [];

    // Load external images
    const externalImages = {};
    try {
      const files = await fs.readdir(EXTERNAL_IMAGE_DIR);
      files.forEach((file) => {
        if (file.match(/\.(png|jpg|jpeg)$/i)) {
          externalImages[file.split(".")[0]] = file;
        }
      });
      console.log(
        `Found ${
          Object.keys(externalImages).length
        } external images in ${EXTERNAL_IMAGE_DIR}`
      );
    } catch (e) {
      console.log(`⚠️ Failed to read external image directory: ${e.message}`);
    }

    for (const worksheet of workbook.worksheets) {
      let currentCategory = "";
      const imgCellMap = {};

      // Map images (exceljs doesn't directly support image anchors; use external images as primary)
      console.log(`Processing sheet: ${worksheet.name}`);

      for (let rowNum = 1; rowNum <= worksheet.rowCount; rowNum++) {
        const row = worksheet.getRow(rowNum);
        const rowValues = row.values || [];
        console.log(`Row ${rowNum} values: ${JSON.stringify(rowValues)}`);

        if (!rowValues.some((val) => val)) {
          console.log(`Skipping empty row ${rowNum}`);
          continue;
        }

        // Skip header rows
        if (
          rowValues.some(
            (val) =>
              typeof val === "string" &&
              ["NAME", "CODE", "PRICE"].includes(val.toUpperCase())
          )
        ) {
          console.log(
            `Skipping header row ${rowNum}: ${JSON.stringify(rowValues)}`
          );
          continue;
        }

        // Category detection
        if (
          rowValues[2] &&
          typeof rowValues[2] === "string" &&
          ["SHOWER TOILET", "E-BIDET", "TOILETS", "BIDET"].some((keyword) =>
            rowValues[2].toUpperCase().includes(keyword)
          )
        ) {
          currentCategory = rowValues[2];
          console.log(`Detected category in row ${rowNum}: ${currentCategory}`);
          continue;
        }

        // Extract data (columns B, C, D => index 2, 3, 4)
        const name = rowValues[2];
        const code = rowValues[3];
        const price = rowValues[4];

        if (!name || !code) {
          console.log(`Skipping row ${rowNum}: Missing name or code`);
          continue;
        }

        // Clean and format price
        const cleanedPrice = cleanPrice(price);
        const formattedPrice =
          typeof cleanedPrice === "number"
            ? `₹ ${cleanedPrice.toLocaleString("en-IN")}`
            : price;

        // Handle image
        let imagePath = null;

        // Try embedded image (simplified; exceljs image handling is limited)
        const key = `${rowNum}_1`;
        if (imageMap[Object.keys(imageMap)[rowNum - 1]]) {
          console.log(`Found embedded image for code ${code} at row ${rowNum}`);
          imagePath = await saveImage(
            imageMap[Object.keys(imageMap)[rowNum - 1]],
            code,
            OUTPUT_IMAGE_DIR
          );
        } else {
          console.log(
            `⚠️ No embedded image found for code ${code} at row ${rowNum}`
          );
        }

        // Try external image
        if (!imagePath) {
          const sanitizedCode = code
            .replace(/”/g, "")
            .replace(/ /g, "_")
            .replace(/\//g, "_");
          for (const variant of [code, sanitizedCode]) {
            if (externalImages[variant]) {
              const externalImgPath = path.join(
                EXTERNAL_IMAGE_DIR,
                externalImages[variant]
              );
              console.log(
                `Found external image for code ${code}: ${externalImgPath}`
              );
              const imgData = await fs.readFile(externalImgPath);
              imagePath = await saveImage(imgData, code, OUTPUT_IMAGE_DIR);
              break;
            }
          }
        }

        // Log column A content
        if (rowValues[1] != null) {
          console.log(`Column A content for code ${code}: ${rowValues[1]}`);
          if (rowValues[1] === "#VALUE!") {
            console.log(
              `⚠️ Column A contains #VALUE! for code ${code}, possibly due to a formula or protection`
            );
          }
        }

        // Push data
        allData.push({
          sheet: worksheet.name,
          category: currentCategory,
          name,
          code,
          price: formattedPrice,
          image_path: imagePath,
        });
      }
    }

    // Write JSON
    await fs.writeFile(
      OUTPUT_JSON_PATH,
      JSON.stringify(allData, null, 2),
      "utf-8"
    );
    console.log(
      `✅ Done. Extracted ${allData.length} items to ${OUTPUT_JSON_PATH}`
    );
  } catch (e) {
    console.error(`Fatal error: ${e.message}`);
    process.exit(1);
  }
}

main();

File: backend/data/extract.py
import os
import json
from openpyxl import load_workbook
from openpyxl.utils import column_index_from_string
from PIL import Image
from io import BytesIO
from zipfile import ZipFile

# === Config ===
EXCEL_PATH = "OnePager New MRP.xlsx"
OUTPUT_IMAGE_DIR = "img"
OUTPUT_JSON_PATH = "output.json"
IMAGE_COLUMN = 'A'  # Images are in column A

# Validate Excel file
if not os.path.exists(EXCEL_PATH):
    raise FileNotFoundError(f"Excel file not found: {EXCEL_PATH}")

# Create output directory and verify permissions
try:
    os.makedirs(OUTPUT_IMAGE_DIR, exist_ok=True)
    # Test write permission by creating a temporary file
    test_file = os.path.join(OUTPUT_IMAGE_DIR, "test.txt")
    with open(test_file, 'w') as f:
        f.write("test")
    os.remove(test_file)
except PermissionError:
    raise PermissionError(f"Cannot write to directory: {OUTPUT_IMAGE_DIR}")

def extract_images_from_xlsx(xlsx_path):
    image_map = {}
    try:
        with ZipFile(xlsx_path, 'r') as zip_ref:
            image_files = [f for f in zip_ref.namelist() if f.startswith('xl/media/')]
            for idx, img_file in enumerate(image_files):
                with zip_ref.open(img_file) as f:
                    image_map[idx + 1] = f.read()
        print(f"Extracted {len(image_map)} images from Excel")
    except Exception as e:
        print(f"⚠️ Failed to extract images from Excel: {e}")
    return image_map

def get_image_key(img):
    anchor = img.anchor
    if hasattr(anchor, 'from_'):
        row = anchor.from_.row + 1
        col = anchor.from_.col + 1
        return f"{row}_{col}"
    return None

def clean_price(price):
    """Clean price string (e.g., '₹ 4,72,950.00') to a float."""
    if isinstance(price, str):
        price = price.replace('₹', '').replace(',', '').strip()
        try:
            return float(price)
        except ValueError:
            return None
    return price

def main():
    wb = load_workbook(EXCEL_PATH, data_only=True)
    image_map = extract_images_from_xlsx(EXCEL_PATH)
    all_data = []

    for sheet in wb.worksheets:
        current_category = ""
        img_cell_map = {}

        # Step 1: Map images to cell positions
        for idx, img in enumerate(sheet._images):
            key = get_image_key(img)
            if key:
                img_cell_map[key] = idx + 1
        print(f"Image cell map for sheet {sheet.title}: {len(img_cell_map)} mappings")

        for row in sheet.iter_rows(min_row=1, values_only=False):
            row_num = row[0].row
            row_values = [
                cell.value.strip() if isinstance(cell.value, str) else cell.value
                for cell in row
            ]

            if not any(row_values):
                continue

            # Step 2: Skip header rows
            if any(
                isinstance(val, str) and val.upper() in ["NAME", "CODE", "PRICE"]
                for val in row_values
            ):
                print(f"Skipping header row {row_num}: {row_values}")
                continue

            # Step 3: Category detection
            if (
                len(row_values) >= 2
                and isinstance(row_values[1], str)
                and any(
                    keyword in row_values[1].upper()
                    for keyword in ["SHOWER TOILET", "E-BIDET", "TOILETS", "BIDET"]
                )
            ):
                current_category = row_values[1]
                print(f"Detected category in row {row_num}: {current_category}")
                continue

            # Step 4: Extract name/code/price (cols B, C, D => index 1, 2, 3)
            try:
                name = row_values[1]  # Column B
                code = row_values[2]  # Column C
                price = row_values[3]  # Column D
            except IndexError:
                print(f"Skipping row {row_num}: Insufficient columns")
                continue

            if not name or not code:
                print(f"Skipping row {row_num}: Missing name or code")
                continue

            # Clean and format price
            cleaned_price = clean_price(price)
            formatted_price = (
                f"₹ {int(cleaned_price):,}" if isinstance(cleaned_price, (int, float)) else price
            )

            # Step 5: Handle image
            col_index = column_index_from_string(IMAGE_COLUMN)
            key = f"{row_num}_{col_index}"
            image = next((i for k, i in img_cell_map.items() if k == key), None)
            image_path = None

            if image and image in image_map:
                try:
                    img_bytes = image_map[image]
                    if not img_bytes or len(img_bytes) < 100:  # Basic validation
                        print(f"⚠️ Invalid image data for code {code} (row: {row_num})")
                    else:
                        img = Image.open(BytesIO(img_bytes))
                        ext = img.format.lower() if img.format else "png"
                        filename = f"{code}.{ext}"
                        filepath = os.path.join(OUTPUT_IMAGE_DIR, filename)
                        img.save(filepath, quality=95)  # Ensure high quality
                        if os.path.exists(filepath):
                            image_path = f"./img/{filename}"
                            print(f"Saved image for code {code}: {image_path}")
                        else:
                            print(f"⚠️ Failed to verify saved image for code {code}")
                except (Image.UnidentifiedImageError, OSError) as e:
                    print(f"⚠️ Invalid image format for code {code}: {e}")
                except PermissionError as e:
                    print(f"⚠️ Permission error saving image for code {code}: {e}")
                except Exception as e:
                    print(f"⚠️ Failed to save image for code {code}: {e}")
            else:
                print(f"⚠️ Image not found for code {code} (row: {row_num})")
                image_path = "./img/placeholder.png"

            # Step 6: Push data to final array
            all_data.append({
                "sheet": sheet.title,
                "category": current_category,
                "name": name,
                "code": code,
                "price": formatted_price,
                "image_path": image_path
            })

    # Write to JSON
    try:
        with open(OUTPUT_JSON_PATH, 'w', encoding='utf-8') as f:
            json.dump(all_data, f, indent=2, ensure_ascii=False)
        print(f"✅ Done. Extracted {len(all_data)} items to {OUTPUT_JSON_PATH}")
    except Exception as e:
        print(f"⚠️ Failed to write JSON file: {e}")

if __name__ == "__main__":
    main()